src/v/kafka/server/replicated_partition.cc

```diff
@@ -205,4 +207,46 @@ std::optional<model::offset> replicated_partition::get_leader_epoch_last_offset(
     }
     return std::nullopt;
 }
+
+ss::future<bool> replicated_partition::is_fetch_offset_valid(
+  model::offset fetch_offset, model::timeout_clock::time_point deadline) {
+    /**
+     * Make sure that we will update high watermark offset if it isn't yet
+     * initialized.
+     *
+     * We use simple heuristic here to determine if high watermark update is
+     * required.
+     *
+     * We only request update if fetch_offset is in range beteween current high
+     * water mark and log end i.e. if high watermark is updated consumer will
+     * receive data. This way we will only update high watermark when leadership
+     * changed after partition is moved or redpanda restarted.
```

Comment:
```
This way we will only update high watermark when leadership changed after partition is moved or redpanda restarted.
```
it's a bit unclear how exactly offsets are used to detect partition move / restart. in some ways it sounds like this would be based on tracking terms. are these offsets some how implicitly communicating a property about terms?

Reply:
Offsets aren't used to detect the change. Offsets are reset to initial values as a result of partition move.